crates/matrix-sdk-crypto/src/olm/account.rs

```diff
@@ -1471,10 +1471,22 @@ impl Account {
             // Ed25519 key of the sender until we decrypt room events. This
             // ensures that we receive the room key even if we don't have access
             // to the device.
-            if !matches!(*event, AnyDecryptedOlmEvent::RoomKey(_)) {
-                let Some(device) =
+            if !matches!(event.as_ref(), AnyDecryptedOlmEvent::RoomKey(_)) {
+                let device = if let Some(device) =
                     store.get_device_from_curve_key(event.sender(), sender_key).await?
-                else {
+                {
+                    device
+                } else if let AnyDecryptedOlmEvent::RoomKeyBundle(_) = event.as_ref() {
+                    // If this is a room key bundle, and we don't have the device in our store,
+                    // we're requiring the device keys to be part of the `AnyDecryptedOlmEvent`.
+                    //
+                    // Othrwise we'll throw an error refusing to decrypt the room key bundle.
+                    let device_keys =
+                        event.sender_device_keys().ok_or(EventError::MissingSigningKey)?;
+                    let device_data = DeviceData::try_from(device_keys).unwrap();
+
+                    store.wrap_device_data(device_data).await?
```

Comment:
I think/hope/believe that `Self::check_sender_device_keys` has already done the check on the Ed25519 key, so there's no point in wrapping the `sender_device_keys` as a Device and doing it again.

Also, the level of indentation is getting painful here.

Could we write this as:

```rust
if let AnyDecryptedOlmEvent::RoomKeyBundle(_) = event.as_ref() {
    // We require sender device keys for room key bundle events.
    event.sender_device_keys().ok_or(EventError::MissingSigningKey)?;
} else if matches!(event.as_ref(), AnyDecryptedOlmEvent::RoomKey(_)) {
   // no-op
} else {
   // ... existing code
}
```

... but with more comments?

(why are we using different syntaxes for matching `RoomKey` and `RoomKeyBundle`? Let's clean that up too, if possible)

Reply:
Hmm, but this is not quite the same. In your snippet we unconditionally require `sender_device_keys` for room key bundles, while in my patch we only require them if no device is found in the store. That is, we fall back to the `sender_device_keys`.

I guess that's fine?

Though if we do the fallback to `sender_device_keys` method, we'll be able to more easily migrate to this logic for all other to-device message types.

Reply:
I think if we're going to require `sender_device_keys` sometimes, we might as well require it all the time. We don't really want the sender trying to second-guess whether to include them. 

Further, if there's a future client impl that *doesn't* put `sender_device_keys` in the key bundle event, and we only require them if we don't know about the device, then that's going to work 99% of the time, which is the worst possible situation for debugging.

Reply:
Alright, that's a fair point.

Reply:
Ok, refactored all of that in https://github.com/matrix-org/matrix-rust-sdk/pull/5118/commits/22b7377626fc8980ca37ee77f23286e85e95e3c1.